### ISSUE TITLE
Overridden assets do not always resolve to the correct Sprinkle in uf-bundle-build

### ISSUE THREAD

**Issue by alexweissman** [userfrosting/UserFrosting#661]
Separate but possibly related to #660 (or not).  It would seem that overridden assets do not always resolve to the correct Sprinkle.  I have the following three Sprinkles:

- `core`
- `nuttermans`
- `site`

The `core` Sprinkle contains the following asset:
`sprinkles/core/vendor/select2/dist/js/select2.full.js`

But then I have a different version of this same asset in `nuttermans`:
`sprinkles/nuttermans/vendor/select2/dist/js/select2.full.js`

When using raw assets, it correctly loads the overriding version in `nuttermans`.  However when I run `uf-bundle-build`, it copies the wrong version of this asset to `public/` (the `core` version instead of the `nuttermans` version).

What's even stranger is that, if I put the asset in the `site` Sprinkle, then `uf-bundle-build` _does_ correctly load the asset from `site`.

Is there a possible race condition going on here, seeing as how gulp tasks execute asynchronously?

**Comment by alexweissman** [comment]
This is a real problem - I have a project where even the _favicons_ aren't being overridden properly, and it's pulling the default UF favicons instead of my Sprinkle's favicons.

**Comment by Silic0nS0ldier** [comment]
What collision condition is being used? (and is this in dev or production)

**Comment by Silic0nS0ldier** [comment]
Oh, and in regards to the 'race condition', that'd be no. I had to specially design everything to prevent that. (its the reason its clunky).

Long term solution is a complete overhaul of asset building system, but for now we should try to plug the hole. (once I know more)

**Comment by alexweissman** [comment]
Are you sure that your special design is actually forcing a correct order, though?

```
    // Gulp v4 src respects order, until it is released, use this alternative.
    return plugins.srcOrderedGlobs(sprinkleAssets)
            .pipe(plugins.copy('../public/assets/', {prefix: 5}));// And gulp.dest doesn't give us the control needed.
```

I thought `pipe` was by its very nature, an asynchronous command.

**Comment by Silic0nS0ldier** [comment]
If there was a race condition, its now a non-issue. The npm frontend assets work has replaced the suspected asynchronous code with fully synchronous code.

**Comment by Silic0nS0ldier** [comment]
Just for completeness, 4.2 the suspect region completely revised as part of the assets overhaul. Unsure if the previous fix has landed in a 4.1.x release however.

**Comment by Silic0nS0ldier** [comment]
The original cause was likely down to a race condition in the copy command queuing up multiple copy actions at once since it uses native code which doesn't have the single threaded constraint of JS (things can happen in parallel).

@alexweissman Regarding your question about `pipe` being asynchronous, it isn't. Rather it delays work until the main event loop is clear (just like `Promise`s) to create the illusion of asynchronous code. If fetch was completely implemented in the JS thread for example, you would see the UI lock up a hell of a lot more across websites. A few heavyweight node tools use Web Workers to achieve true asynchronous code (it uses real threads), but it isn't more widely adopted because it does take some work to get set up (thankfully JS tooling is started to be developed to make them easier to work).

Anyway, this is resolved in `4.2-beta.1`.
